### PR TITLE
Remove map blockers, parent spawnpoints, detect map end

### PR DESCRIPTION
**Pull request by luttje**
As discussed in issues #1, some \~creative~ solutions

This also fixes health stations which use trigger_hurt with -200 (this doesn't do anything in gmod?)

It sets the damage of those triggers to 0 and marks them with HealAmount. They are healed when the EntityTakeDamage hook notices damage == 0.

If a HealAmount of 200 is followed by a trigger_hurt of 100 damage (read from keyvalues) within END_CHECK_INTERVAL (2) seconds, that is classified as ending the map. A hook is called `SlidePlayerFinished(ply, finishedAt)`

### CODE REVIEW

**Comment by luttje** [comment]
With 09209cd68bf8c93e6f771941c7b81ce4de04e360 I'm making it so my other minigames don't have terrible physics after slide has mangled the server settings

**Comment by luttje** [comment]
Also found that `slide_city_remake.bsp` ends in a teleport and not the -200 100 hurt sequence. LAME! So I added a hook that checks whether the player teleported to the teleport at the end (by name)

Also added a basic round system that cleans up the map after everyone dies and prevents respawning.

File: gamemode/mapfixes/triggers.lua [hunk 1]
```diff
@@ -1,5 +1,5 @@
 --[[
-	Slide - gamemode/mapfixes/triggers.lua
+  Slide - gamemode/mapfixes/triggers.lua
```
Comment (Lexicality) [inline gamemode/mapfixes/triggers.lua:2]: All the files in this project use tabs, can you please install [an editorconfig plugin](https://editorconfig.org/#download) for your editor and reformat these files

File: gamemode/init.lua [hunk 2]
```diff
@@ -88,6 +112,15 @@ function GM:PlayerSilentDeath(ply)
 	end
 end
 
+--- Do not allow respawning by pressing space or clicking unless we only played last round
+function GM:PlayerDeathThink(ply)
+	if(ply.RoundNumber and ply.RoundNumber == self.Round)then
```
Comment (Lexicality) [inline gamemode/init.lua:117]: Also, I used [LuaFormatter](https://github.com/Koihik/LuaFormatter#editor-support) for consistency, please can you format everything with it

File: gamemode/mapfixes/triggers.lua [hunk 3]
```diff
@@ -17,28 +17,107 @@
...
+end
+
+--- from: https://wiki.facepunch.com/gmod/ENTITY:TriggerOutput
+function GM:SetupTeleportHook()
+	MapLua = IsValid(MapLua) and MapLua or ents.Create("lua_run")
+	MapLua:SetName("triggerhook")
+	MapLua:Spawn()
+
+	for _, teleport in pairs(ents.FindByClass("trigger_teleport"))do
+		teleport:Fire("AddOutput", "OnStartTouch triggerhook:RunPassedCode:hook.Run('RawOnTeleport'):0:-1")
```
Comment (Lexicality) [inline gamemode/mapfixes/triggers.lua:56]: Any particular reason to write the code each time rather than setting the `Code` keyvalue?

Comment (Lexicality) [reply gamemode/mapfixes/triggers.lua:56]: Also why do you use an intermediary hook?

File: gamemode/rounds.lua [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    if(IsValid(ply) and ply:Alive())then
+      ply:KillSilent()
+    end
+  end
+
+  -- Do this in a timer, so we don't remove an invalid entity (like the skull that spawns on death, or ragdoll??)
+  -- Prevents a crash by doing game.CleanUpMap in a timer, unsure why
+  timer.Simple(0, function()
+    -- Quick workaround for issue @ https://github.com/Facepunch/garrysmod-issues/issues/3637
+    game.CleanUpMap(false, {"env_fire", "entityflame", "_firesmoke"})
```
Comment (Lexicality) [inline gamemode/rounds.lua:35]: There are very few maps that actually need this called on them regularly, just ones like collapse. A couple have destructible bits that you only really want to clean up every 5-10 minutes and a couple have one time doors that can be mapfixed to re-open after 30 seconds.

That said, I'm not sure even collapse needs this called every single round.

**Review by Lexicality — COMMENTED** [review]
initial thoughts, more later

File: gamemode/mapfixes/triggers.lua (hunk 3 above)
Comment (luttje) [reply gamemode/mapfixes/triggers.lua:56]: I don't understand your first comment, what do you mean with writing the code each time and which Code keyvalue?

Comment (luttje) [reply gamemode/mapfixes/triggers.lua:56]: The intermediary hook is so the "higher level" hook doesn't have to access globals (ACTIVATOR, CALLER). Basically so we can use the hook in a more conventional gmod way.

Comment (Lexicality) [reply gamemode/mapfixes/triggers.lua:56]: eg
```lua
	MapLua = IsValid(MapLua) and MapLua or ents.Create("lua_run")
	MapLua:SetName("triggerhook")
	MapLua:SetKeyValue(
		"Code", "gamemode.Call('OnPlayerTeleport', ACTIVATOR, CALLER)"
	)
	MapLua:Spawn()

	for _, teleport in pairs(ents.FindByClass("trigger_teleport")) do
		teleport:Fire("AddOutput", "OnStartTouch triggerhook:RunCode::0:-1")
	end
```

Comment (luttje) [reply gamemode/mapfixes/triggers.lua:56]: Oh that's cool, didn't know that was possible. I got that code from here: https://wiki.facepunch.com/gmod/ENTITY:TriggerOutput

Comment (Lexicality) [reply gamemode/mapfixes/triggers.lua:56]: Sadly the state of gmod and the wiki is you generally shouldn't trust any example there without going and checking out the source first 🙈

**Comment by luttje** [comment]
Now that you've taken the good commits and revised it, I guess this PR can be closed. I'll leave the honours to you

**Comment by Lexicality** [comment]
I wanted to leave it open until I've got the respawn logic out, but can close if you'd prefer

**Pull request closed without merging**